pkg/offsetguess/offsetguess.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2017 Kinvolk GmbH
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build linux
+
+package offsetguess
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+	"unsafe"
+
+	"github.com/iovisor/gobpf/elf"
+
+	"github.com/weaveworks/tcptracer-bpf/pkg/byteorder"
+)
+
+type tcpTracerState uint64
+
+const (
+	uninitialized tcpTracerState = iota
+	checking
+	checked
+	ready
+)
+
+type guessWhat uint64
+
+const (
+	guessSaddr guessWhat = iota
+	guessDaddr
+	guessFamily
+	guessSport
+	guessDport
+	guessNetns
+	guessDaddrIPv6
+)
+
+type tcpTracerStatus struct {
+	status          tcpTracerState
+	pidTgid         uint64
+	what            guessWhat
+	offsetSaddr     uint64
+	offsetDaddr     uint64
+	offsetSport     uint64
+	offsetDport     uint64
+	offsetNetns     uint64
+	offsetIno       uint64
+	offsetFamily    uint64
+	offsetDaddrIPv6 uint64
+	err             byte
+	saddr           uint32
+	daddr           uint32
+	sport           uint16
+	dport           uint16
+	netns           uint32
+	family          uint16
+	daddrIPv6       [4]uint32
+}
+
+func listen(url, netType string, finish chan struct{}) {
+	l, err := net.Listen(netType, url)
+	if err != nil {
+		panic(err)
+	}
+	select {
+	case <-finish:
+		l.Close()
+		return
+	}
+}
+
+func compareIPv6(a, b [4]uint32) bool {
+	for i := 0; i < 4; i++ {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func ownNetNS() (uint64, error) {
+	var s syscall.Stat_t
+	if err := syscall.Stat("/proc/self/ns/net", &s); err != nil {
+		return 0, err
+	}
+	return s.Ino, nil
+}
+
+func ipFromUint32Arr(ipv6Addr [4]uint32) net.IP {
+	buf := make([]byte, 16)
+	for i := 0; i < 16; i++ {
+		buf[i] = *(*byte)(unsafe.Pointer((uintptr(unsafe.Pointer(&ipv6Addr[0])) + uintptr(i))))
+	}
+	return net.IP(buf)
+}
+
+func htons(a uint16) uint16 {
+	arr := make([]byte, 2)
+	binary.BigEndian.PutUint16(arr, a)
+	return byteorder.Host.Uint16(arr)
```

Comment:
Either way, you are using `unsafe.Pointer()` in host. We are only using Host once, so it may be worth doing the unsafe cast/bit operations here and ditching the `byteorder` package.